File: src/test/java/com/cp/csms/authentication/AuthenticationConsumerTest.java
package com.cp.csms.authentication;

import com.cp.csms.authentication.validation.ValidationService;
import com.cp.csms.common.AuthenticationMessage;
import com.cp.csms.common.AuthenticationResponse;
import com.cp.csms.common.AuthenticationStatus;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class AuthenticationConsumerTest {

    @Mock
    private ValidationService validationService;

    @Mock
    private AuthenticationService authenticationService;

    @InjectMocks
    private AuthenticationConsumer consumer;

    @Test
    void shouldReturnInvalidWhenValidationFails() {
        final AuthenticationMessage message = new AuthenticationMessage("req-123", "invalid-token");
        when(validationService.isValid(message)).thenReturn(false);

        final AuthenticationResponse response = consumer.handleAuthRequest(message);

        assertThat(response.getRequestId()).isEqualTo("req-123");
        assertThat(response.getAuthenticationStatus()).isEqualTo(AuthenticationStatus.INVALID);
        verify(validationService).isValid(message);
        verifyNoInteractions(authenticationService);
    }

    @Test
    void shouldReturnAcceptedWhenValidationPassesAndAuthenticationSucceeds() {
        final AuthenticationMessage message = new AuthenticationMessage("req-456", "valid-token");
        when(validationService.isValid(message)).thenReturn(true);
        when(authenticationService.authenticate("valid-token")).thenReturn(AuthenticationStatus.ACCEPTED);

        final AuthenticationResponse response = consumer.handleAuthRequest(message);

        assertThat(response.getRequestId()).isEqualTo("req-456");
        assertThat(response.getAuthenticationStatus()).isEqualTo(AuthenticationStatus.ACCEPTED);
        verify(validationService).isValid(message);
        verify(authenticationService).authenticate("valid-token");
    }

    @Test
    void shouldReturnRejectedWhenValidationPassesButTokenIsDisabled() {
        final AuthenticationMessage message = new AuthenticationMessage("req-789", "disabled-token");
        when(validationService.isValid(message)).thenReturn(true);
        when(authenticationService.authenticate("disabled-token")).thenReturn(AuthenticationStatus.REJECTED);

        final AuthenticationResponse response = consumer.handleAuthRequest(message);

        assertThat(response.getRequestId()).isEqualTo("req-789");
        assertThat(response.getAuthenticationStatus()).isEqualTo(AuthenticationStatus.REJECTED);
        verify(validationService).isValid(message);
        verify(authenticationService).authenticate("disabled-token");
    }

    @Test
    void shouldReturnUnknownWhenValidationPassesButTokenIsNotFound() {
        final AuthenticationMessage message = new AuthenticationMessage("req-999", "unknown-token");
        when(validationService.isValid(message)).thenReturn(true);
        when(authenticationService.authenticate("unknown-token")).thenReturn(AuthenticationStatus.UNKNOWN);

        final AuthenticationResponse response = consumer.handleAuthRequest(message);

        assertThat(response.getRequestId()).isEqualTo("req-999");
        assertThat(response.getAuthenticationStatus()).isEqualTo(AuthenticationStatus.UNKNOWN);
        verify(validationService).isValid(message);
        verify(authenticationService).authenticate("unknown-token");
    }

    @Test
    void shouldPreserveRequestIdInResponse() {
        final String requestId = "unique-request-id-12345";
        final AuthenticationMessage message = new AuthenticationMessage(requestId, "some-token");
        when(validationService.isValid(message)).thenReturn(false);

        final AuthenticationResponse response = consumer.handleAuthRequest(message);

        assertThat(response.getRequestId()).isEqualTo(requestId);
    }
}

File: src/test/kotlin/com/cp/csms/authentication/AuthenticationConsumerTest.kt
package com.cp.csms.authentication

import com.cp.csms.authentication.validation.ValidationService
import com.cp.csms.common.AuthenticationMessage
import com.cp.csms.common.AuthenticationStatus
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoInteractions
import org.mockito.kotlin.whenever

@ExtendWith(MockitoExtension::class)
class AuthenticationConsumerTest {

    @Mock
    private lateinit var validationService: ValidationService

    @Mock
    private lateinit var authenticationService: AuthenticationService

    @InjectMocks
    private lateinit var consumer: AuthenticationConsumer

    @Test
    fun `should return invalid when validation fails`() {
        val message = AuthenticationMessage("req-123", "invalid-token")
        whenever(validationService.isValid(message)).thenReturn(false)

        val response = consumer.handleAuthRequest(message)

        assertThat(response.requestId).isEqualTo("req-123")
        assertThat(response.authenticationStatus).isEqualTo(AuthenticationStatus.INVALID)
        verify(validationService).isValid(message)
        verifyNoInteractions(authenticationService)
    }

    @Test
    fun `should return accepted when validation passes and authentication succeeds`() {
        val message = AuthenticationMessage("req-456", "valid-token")
        whenever(validationService.isValid(message)).thenReturn(true)
        whenever(authenticationService.authenticate("valid-token")).thenReturn(AuthenticationStatus.ACCEPTED)

        val response = consumer.handleAuthRequest(message)

        assertThat(response.requestId).isEqualTo("req-456")
        assertThat(response.authenticationStatus).isEqualTo(AuthenticationStatus.ACCEPTED)
        verify(validationService).isValid(message)
        verify(authenticationService).authenticate("valid-token")
    }

    @Test
    fun `should return rejected when validation passes but token is disabled`() {
        val message = AuthenticationMessage("req-789", "disabled-token")
        whenever(validationService.isValid(message)).thenReturn(true)
        whenever(authenticationService.authenticate("disabled-token")).thenReturn(AuthenticationStatus.REJECTED)

        val response = consumer.handleAuthRequest(message)

        assertThat(response.requestId).isEqualTo("req-789")
        assertThat(response.authenticationStatus).isEqualTo(AuthenticationStatus.REJECTED)
        verify(validationService).isValid(message)
        verify(authenticationService).authenticate("disabled-token")
    }

    @Test
    fun `should return unknown when validation passes but token is not found`() {
        val message = AuthenticationMessage("req-999", "unknown-token")
        whenever(validationService.isValid(message)).thenReturn(true)
        whenever(authenticationService.authenticate("unknown-token")).thenReturn(AuthenticationStatus.UNKNOWN)

        val response = consumer.handleAuthRequest(message)

        assertThat(response.requestId).isEqualTo("req-999")
        assertThat(response.authenticationStatus).isEqualTo(AuthenticationStatus.UNKNOWN)
        verify(validationService).isValid(message)
        verify(authenticationService).authenticate("unknown-token")
    }

    @Test
    fun `should preserve request id in response`() {
        val requestId = "unique-request-id-12345"
        val message = AuthenticationMessage(requestId, "some-token")
        whenever(validationService.isValid(message)).thenReturn(false)

        val response = consumer.handleAuthRequest(message)

        assertThat(response.requestId).isEqualTo(requestId)
    }
}
